Don't copy DynamicType argument to DynamicType::create

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151850
* #151849
* #151810
* #151807
* #151806
* #151805
* #151804
* #151803
* #151802
* __->__ #151801

This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

swolchok committed:
Don't copy DynamicType argument to DynamicType::create

This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

[ghstack-poisoned]

swolchok committed:
Update on "Don't copy DynamicType argument to DynamicType::create"


This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

[ghstack-poisoned]

swolchok committed:
Update on "Don't copy DynamicType argument to DynamicType::create"


This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

[ghstack-poisoned]

Skylion007 commented on aten/src/ATen/core/dynamic_type.cpp:
```diff
@@ -78,15 +78,13 @@ DynamicType::~DynamicType() {
-    return dyn;
+    return SingletonTypePtr<const DynamicType>(dynRaw);
   }
   return std::shared_ptr<const DynamicType>(new DynamicType{other});
```
Why isn't this using make_shared? It is doing 2 heap allocations with this code unnecessarily.

swolchok replied:
I did try fixing this, and I was reminded that you can't use make_shared if your constructor is private.

swolchok committed:
rebase and minor fixes as requested on "Don't copy DynamicType argument to DynamicType::create"


This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

[ghstack-poisoned]

swolchok commented:
ping reviewers @zhxchen17 @malfet @seemethere @atalman

malfet commented on aten/src/ATen/core/dynamic_type.cpp:
```diff
@@ -78,15 +78,13 @@ DynamicType::~DynamicType() {
-  }
-  if (auto dyn = other.cast<DynamicType>()) {
-    return dyn;
+    return SingletonTypePtr<const DynamicType>(dynRaw);
```
Can one have some sort of static assert that guarantees that copy constructor is not invoked when casting `SingletonTypePtr` to `SingletonOrSharedTypePtr`

swolchok replied:
making that conversion cheap is a core goal of SingletonOrSharedTypePtr. If somebody changes that they're breaking everything it's for. https://github.com/pytorch/pytorch/blob/4d2d833976a3616953aee3fb75892fc0c7988309/aten/src/ATen/core/jit_type_base.h#L187-L188

malfet replied:
Sure, so it would be nice to have static assert somewhere, to prevent someone accidentally breaking it

swolchok committed:
Update on "Don't copy DynamicType argument to DynamicType::create"


This improves performance of DynamicType::isSubtypeOfExt.

Differential Revision: [D73129449](https://our.internmc.facebook.com/intern/diff/D73129449/)

[ghstack-poisoned]
